### PR TITLE
Add core22 colcon plugin / ros2-humble extension

### PR DESCRIPTION
**Pull request by artivis**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

Add the `colcon` plugin for `core22` together with the `ros2-humble` extension.

~~This PR is mostly ready, at the time of writing it still requires a way to enable experimental extensions with providers (e.g. lxd).~~
Note, not all spread tests were ported from `core20`. I plan on revisiting the suite later on.

Depends on, https://github.com/snapcore/snapcraft/pull/3741, https://github.com/snapcore/snapcraft/pull/3758, https://github.com/snapcore/snapcraft/pull/3775, https://github.com/canonical/craft-parts/pull/234

### CODE REVIEW

File: snapcraft/parts/plugins/_ros.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                    '"${ROS_VERSION}"',
+                    "--ros-distro",
+                    '"${ROS_DISTRO}"',
+                    "--target-arch",
+                    '"${CRAFT_TARGET_ARCH}"',
+                    "--stage-cache-dir",
+                    str(self._part_info.cache_dir.resolve()),
+                    "--base",
+                    "core22",
+                    # f"{self._part_info.base}",
```
Comment (Guillaumebeuzeboc) [inline snapcraft/parts/plugins/_ros.py:170]: Commented line to remove?

File: snapcraft/extensions/ros2_humble.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        ]
+        return {
+            "command-chain": ["snap/command-chain/ros2-launch"],
+            "environment": {
+                "ROS_VERSION": self.ROS_VERSION,
+                "ROS_DISTRO": self.ROS_DISTRO,
+                "PYTHONPATH": ":".join(python_paths),
+                # Various ROS 2 tools (e.g. spdlog logger) keep a cache or a log,
+                # and use $ROS_HOME to determine where to put them.
+                "ROS_HOME": "$SNAP_USER_DATA/ros",
```
[the remark this answers is not in the document]
Comment (artivis) [reply snapcraft/extensions/ros2_humble.py:83]: We've used `ros` in [legacy](https://github.com/snapcore/snapcraft/blob/0e9e8cb99eaa7f8da02b0ba3b6a2e467d3c4f643/snapcraft_legacy/plugins/v1/catkin.py#L360), that's why I stuck to it.

File: snapcraft/parts/plugins/_ros.py (hunk 1 above)
Comment (artivis) [reply snapcraft/parts/plugins/_ros.py:170]: That's a 'todo' really...

**Review by Guillaumebeuzeboc — APPROVED** [review]

File: snapcraft/parts/plugins/_ros.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/parts/plugins/_ros.py:170]: what's the TODO?

File: schema/snapcraft.json [hunk 3]
```diff
@@ -912,7 +912,8 @@
                                     "gnome-3-38",
                                     "kde-neon",
                                     "ros1-noetic",
-                                    "ros2-foxy"
+                                    "ros2-foxy",
+                                    "ros2-humble"
```
Comment (sergiusens) [inline schema/snapcraft.json:916]: this file no longer used in new snapcraft

File: snapcraft/parts/plugins/_ros.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+import abc
+import os
+import pathlib
+import re
+import subprocess
+import sys
+from pathlib import Path
+from typing import Dict, List, Set
+
+import click
```
Comment (cmatsuoka) [inline snapcraft/parts/plugins/_ros.py:28]: Ideally we shouldn't use click in non-legacy code.

File: snapcraft/parts/plugins/_ros.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                '--from-paths "${CRAFT_PART_SRC_WORK}"',
+            ]
+            + self._get_build_commands()
+            + self._get_stage_runtime_dependencies_commands()
+        )
+
+
+@click.group()
+def plugin_cli():
+    """Lala."""
```
Comment (cmatsuoka) [inline snapcraft/parts/plugins/_ros.py:193]: This is probably a placeholder?

File: snapcraft/parts/plugins/colcon.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2020 Canonical Ltd
```
Comment (cmatsuoka) [inline snapcraft/parts/plugins/colcon.py:3]: The copyright notice line has a different format.

File: snapcraft/extensions/ros2_humble.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        return {
+            "build-environment": [
+                {"ROS_VERSION": self.ROS_VERSION},
+                {"ROS_DISTRO": self.ROS_DISTRO},
+            ]
+        }
+
+    @overrides
+    def get_parts_snippet(self) -> Dict[str, Any]:
+        source = common.get_extensionsdir()
```
Comment (sergiusens) [inline snapcraft/extensions/ros2_humble.py:98]: this is now in main

**Pull request merged**